vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,416 @@
+/// Use the methods <code>AddCodeLens</code< and <code> RemoveCodeLens</code>
+type CodeLensGeneralTagger (view, buffer) as self =
+    inherit CodeLensDisplayService(view, buffer)
+
+    /// Saves the ui context to switch context for ui related work.
+    let uiContext = SynchronizationContext.Current
+
+    /// The tags changed event to notify if the data for the tags has changed.
+    let tagsChangedEvent = new Event<EventHandler<SnapshotSpanEventArgs>,SnapshotSpanEventArgs>()
+
+    // Tracks the created ui elements per TrackingSpan
+    let uiElements = Dictionary<_,_>()
+    /// Caches the current used trackingSpans per line. One line can contain multiple trackingSpans
+    let mutable trackingSpans = Dictionary<_, Generic.List<_>>()
+    /// Text view for accessing the adornment layer.
+    let mutable view: IWpfTextView = view
+    /// The code lens layer for adding and removing adornments.
+    let mutable codeLensLayer: IAdornmentLayer = view.GetAdornmentLayer "CodeLens"
+    /// Tracks the recent first + last visible line numbers for adornment layout logic.
+    let mutable recentFirstVsblLineNmbr, recentLastVsblLineNmbr = 0, 0
+    /// Tracks the adornments on the layer.
+    let mutable addedAdornments = HashSet()
+    /// Cancellation token source for the layout changed event. Needed to abort previous async-work.
+    let mutable layoutChangedCts = new CancellationTokenSource()
+
+    /// Tracks the last used buffer snapshot, should be preferred used in combination with mutex.
+    let mutable currentBufferSnapshot = null
+
+    /// Helper method which returns the start line number of a tracking span
+    let getTrackingSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
+        snapshot.GetLineNumberFromPosition(trackingSpan.GetStartPoint(snapshot).Position)
+
+    /// Helper method which returns the start line number of a tracking span
+    let tryGetTSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
+        let pos = trackingSpan.GetStartPoint(snapshot).Position
+        if snapshot.Length - 1 < pos then None
+        else pos |> snapshot.GetLineNumberFromPosition |> Some
+    
+    let updateTrackingSpansFast (snapshot:ITextSnapshot) lineNumber =
+        if lineNumber |> trackingSpans.ContainsKey then
+            let currentTrackingSpans = Generic.List(trackingSpans.[lineNumber])
+            for trackingSpan in currentTrackingSpans do
+                let newLineOption = tryGetTSpanStartLine snapshot trackingSpan
+                match newLineOption with 
+                | None -> ()
+                | Some newLine ->
+                    if newLine <> lineNumber then
+                        trackingSpans.[lineNumber].Remove(trackingSpan) |> ignore
+                        if trackingSpans.[lineNumber].Count = 0 then
+                            trackingSpans.Remove lineNumber |> ignore
+                        if newLine |> trackingSpans.ContainsKey |> not then
```

Comment:
It's only my opinion, but I think this is harder to read that `if not (trackingSpans.ContainsKey newLine) then`